### PR TITLE
Update: Add an option to `prefer-const` (fixes #5692)

### LINKED ISSUE #5692: conflict between prefer-const and no-use-before-define (2.5.0 regression)
**What version of ESLint are you using?**
2.5.0 (I think my previous version that didn't fail was 2.4.0, but it could have been a different 2.x)
**What parser (default, Babel-ESLint, etc.) are you using?**
Default
**Please show your full configuration:**
```js
{
...
"no-use-before-define": ["error", "nofunc"],
...
 "prefer-const": "error",
}
```
**What did you do? Please include the actual source code causing the issue.**
```js
            let interval;
            function checkForWindow() {
                const newFocussed = BrowserWindow.getFocusedWindow();
                if (newFocussed !== currentlyFocussed) {
                    clearInterval(interval);
                    newFocussed.toggleDevTools();
                }
            }
            interval = setInterval(checkForWindow, msIntervalToCheckForWindow);
```
**What did you expect to happen?**
no warnings
**What actually happened? Please include the actual, raw output from ESLint.**
either
```sh
  76:13  error  'interval' is never reassigned, use 'const' instead  prefer-const
```
or if I change it to const after the interval
```sh
  72:35  error  'interval' was used before it was defined    no-use-before-define
```

I see the rule says this is an error
```js
// the initializer is separated.
let a;
a = 0;
console.log(a);
```
but this is not
```js
let a;
if (true) {
    a = 0;
}
console.log(a);
```
I would suggest that
```js
let a;
// read or write of a in any sub block (e.g. if or function)
console.log(a);
```
is not an error?

### PR DESCRIPTION
**Pull request by mysticatea**
Fixes #5692.

I added an option `ignoreReadingBeforeInit`.

### CODE REVIEW

File: docs/rules/prefer-const.md [hunk 1]
```diff
@@ -148,10 +151,46 @@ let {a, b} = obj;
 a = a + 1;
 ```
 
+### ignoreReadingBeforeInit
```
Comment (nzakas) [inline docs/rules/prefer-const.md]: Maybe change to `ignoreReadBeforeAssign`?

**Comment by nzakas** [comment]
I'm wondering -- should this be the default behavior rather than adding an option?

**Comment by mysticatea** [comment]
We can change these to `const` in most cases since function declarations are hoisted.

```js
let foo;
doSomething(foo); // this `foo` can change to `undefined`.
foo = 1;
```

```js
let foo;
function bar() {
    doSomething(foo);
}
foo = 1; // this assignment can move to the initializer of the declaration.
```

```js
let foo;
function bar() {
    doSomething(foo);
}
foo = setInterval(bar, 100); // this assignment can move to the initializer of the declaration.
```

**Comment by mysticatea** [comment]
```js
let foo;
const bar = () => doSomething(foo);
foo = setInterval(bar, 100); // false positive here.
```

I can set `true` by default to this option.

**Comment by nzakas** [comment]
Hmm, good point. An option makes sense. :+1:. I think the default should be whatever is the current behavior.

**Comment by nzakas** [comment]
@mysticatea is the default value the same as current behavior?

**Comment by mysticatea** [comment]
Ops, I'm sorry for missing this.

I updated new option name.
Yes, I don't change the default behavior.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**